File: claude-bridge.js
const express = require('express');
const fs = require('fs');
const path = require('path');
const app = express();
const port = 3001;

app.use(express.json());

// Arquivos para comunicação com Claude
const pendingMessages = path.join(__dirname, 'pending_messages.json');
const responses = path.join(__dirname, 'claude_responses.json');
const conversationLog = path.join(__dirname, 'conversation_log.json');

// Função para inicializar arquivos
function initializeFiles() {
    if (!fs.existsSync(pendingMessages)) {
        fs.writeFileSync(pendingMessages, JSON.stringify([], null, 2));
    }
    if (!fs.existsSync(responses)) {
        fs.writeFileSync(responses, JSON.stringify({}, null, 2));
    }
    if (!fs.existsSync(conversationLog)) {
        fs.writeFileSync(conversationLog, JSON.stringify({}, null, 2));
    }
}

// Função para adicionar mensagem à fila
function addPendingMessage(senderId, message) {
    const pending = JSON.parse(fs.readFileSync(pendingMessages, 'utf8'));
    const messageData = {
        id: Date.now().toString(),
        senderId,
        message,
        timestamp: new Date().toISOString(),
        status: 'pending'
    };
    pending.push(messageData);
    fs.writeFileSync(pendingMessages, JSON.stringify(pending, null, 2));
    return messageData.id;
}

// Função para buscar resposta do Claude
function getClaudeResponse(messageId) {
    try {
        const responses = JSON.parse(fs.readFileSync(responses, 'utf8'));
        return responses[messageId] || null;
    } catch (error) {
        return null;
    }
}

// Função para marcar mensagem como respondida
function markAsAnswered(messageId) {
    const pending = JSON.parse(fs.readFileSync(pendingMessages, 'utf8'));
    const updated = pending.map(msg => 
        msg.id === messageId ? { ...msg, status: 'answered' } : msg
    );
    fs.writeFileSync(pendingMessages, JSON.stringify(updated, null, 2));
}

// Função para salvar no log de conversas
function saveToConversationLog(senderId, message, reply) {
    let conversations = JSON.parse(fs.readFileSync(conversationLog, 'utf8'));
    
    if (!conversations[senderId]) {
        conversations[senderId] = [];
    }
    
    conversations[senderId].push({
        timestamp: new Date().toISOString(),
        message,
        reply
    });
    
    fs.writeFileSync(conversationLog, JSON.stringify(conversations, null, 2));
}

// Endpoint principal para receber mensagens do WhatsApp
app.post('/api/whatsapp-chat', async (req, res) => {
    const { senderId, message } = req.body;
    
    console.log(`\n🔔 NOVA MENSAGEM RECEBIDA`);
    console.log(`📱 De: ${senderId}`);
    console.log(`💬 Mensagem: "${message}"`);
    console.log(`⏰ Horário: ${new Date().toLocaleString('pt-BR')}`);
    
    // Criar arquivo de mensagem diretamente no sistema
    const messageId = Date.now().toString();
    const messageFile = path.join(__dirname, `message_${messageId}.txt`);
    const responseFile = path.join(__dirname, `response_${messageId}.txt`);
    
    // Escrever mensagem no arquivo para Claude ver
    const messageContent = `=== MENSAGEM WHATSAPP ===
ID: ${messageId}
De: ${senderId}
Horário: ${new Date().toLocaleString('pt-BR')}
Mensagem: "${message}"

=== INSTRUÇÕES PARA CLAUDE ===
Por favor, responda esta mensagem do WhatsApp.
Sua resposta deve ser natural e útil.
Quando terminar, salve sua resposta no arquivo: response_${messageId}.txt

=== CONTEXTO ===
Esta é uma mensagem real de um usuário do WhatsApp.
Responda como se fosse um assistente inteligente.`;

    fs.writeFileSync(messageFile, messageContent);
    console.log(`📄 Mensagem salva em: ${messageFile}`);
    console.log(`⏳ Aguardando Claude responder em: response_${messageId}.txt`);
    
    // Aguardar resposta do Claude no arquivo
    let attempts = 0;
    const maxAttempts = 60; // 60 segundos
    
    const waitForResponse = () => {
        return new Promise((resolve) => {
            const checkResponse = () => {
                if (fs.existsSync(responseFile)) {
                    const response = fs.readFileSync(responseFile, 'utf8').trim();
                    console.log(`✅ Claude respondeu: "${response}"`);
                    
                    // Limpar arquivos temporários
                    fs.unlinkSync(messageFile);
                    fs.unlinkSync(responseFile);
                    
                    saveToConversationLog(senderId, message, response);
                    resolve(response);
                } else if (attempts < maxAttempts) {
                    attempts++;
                    setTimeout(checkResponse, 1000);
                } else {
                    console.log(`⏰ Timeout - enviando resposta padrão`);
                    const defaultResponse = "Desculpe, não consegui processar sua mensagem no momento. Tente novamente em alguns instantes.";
                    
                    // Limpar arquivo de mensagem
                    if (fs.existsSync(messageFile)) {
                        fs.unlinkSync(messageFile);
                    }
                    
                    saveToConversationLog(senderId, message, defaultResponse);
                    resolve(defaultResponse);
                }
            };
            checkResponse();
        });
    };
    
    const reply = await waitForResponse();
    console.log(`📤 Enviando resposta para WhatsApp\n`);
    
    res.json({ reply });
});

// Endpoint para Claude ver mensagens pendentes
app.get('/api/pending-messages', (req, res) => {
    const pending = JSON.parse(fs.readFileSync(pendingMessages, 'utf8'));
    const pendingOnly = pending.filter(msg => msg.status === 'pending');
    res.json(pendingOnly);
});

// Endpoint para Claude enviar respostas
app.post('/api/claude-response', (req, res) => {
    const { messageId, response } = req.body;
    
    const responses = JSON.parse(fs.readFileSync(responses, 'utf8'));
    responses[messageId] = response;
    fs.writeFileSync(responses, JSON.stringify(responses, null, 2));
    
    console.log(`📝 Claude respondeu mensagem ${messageId}: "${response}"`);
    res.json({ success: true });
});

// Endpoint para ver conversas
app.get('/api/conversations', (req, res) => {
    const conversations = JSON.parse(fs.readFileSync(conversationLog, 'utf8'));
    res.json(conversations);
});

// Endpoint de teste
app.get('/api/test', (req, res) => {
    res.json({ 
        status: 'Sistema Claude Bridge ativo!',
        timestamp: new Date().toISOString(),
        message: 'Pronto para receber mensagens do WhatsApp e processar com Claude' 
    });
});

// Inicializar arquivos e servidor
initializeFiles();

app.listen(port, () => {
    console.log(`🤖 Sistema Claude Bridge iniciado na porta ${port}`);
    console.log(`🌐 Teste: http://localhost:${port}/api/test`);
    console.log(`📥 Mensagens pendentes: http://localhost:${port}/api/pending-messages`);
    console.log(`💬 Conversas: http://localhost:${port}/api/conversations`);
    console.log(`📋 INSTRUÇÕES PARA CLAUDE:`);
    console.log(`   1. Monitore: http://localhost:${port}/api/pending-messages`);
    console.log(`   2. Responda via: POST /api/claude-response`);
    console.log(`   3. Formato: {messageId: "123", response: "sua resposta"}`);
    console.log(`⏰ Iniciado em: ${new Date().toLocaleString('pt-BR')}\n`);
});

File: claude_bridge.py
#!/usr/bin/env python3
"""
Claude Bridge - Ponte Python entre WhatsApp e Claude Code
Monitora mensagens e injeta no histórico do Claude automaticamente
"""

import json
import time
import os
from datetime import datetime
from flask import Flask, request, jsonify
import threading

# Configuração
WHATSAPP_MESSAGES_PATH = '/home/user/whatsapp_messages.json'
CLAUDE_RESPONSE_PATH = '/home/user/claude_response.json'
CLAUDE_CONFIG_PATH = '/home/user/.claude.json'

app = Flask(__name__)

class ClaudeBridge:
    def __init__(self):
        self.last_processed_id = None
        self.running = True
        print("🐍 Claude Bridge Python iniciado!")
        print(f"📱 Monitorando: {WHATSAPP_MESSAGES_PATH}")
        print(f"🧠 Injetando em: {CLAUDE_CONFIG_PATH}")
        print(f"⏰ Iniciado em: {datetime.now().strftime('%d/%m/%Y, %H:%M:%S')}")
        print("========================\n")

    def inject_message_to_claude_history(self, sender_id, message, message_id):
        """Injeta mensagem no histórico do Claude Config"""
        try:
            # Ler configuração atual do Claude
            with open(CLAUDE_CONFIG_PATH, 'r', encoding='utf-8') as f:
                claude_config = json.load(f)
            
            # Preparar mensagem formatada para o Claude ver
            whatsapp_message = f"""🔔 MENSAGEM WHATSAPP RECEBIDA!

📱 **De:** {sender_id}
💬 **Mensagem:** "{message}"
🆔 **ID:** {message_id}
⏰ **Horário:** {datetime.now().strftime('%d/%m/%Y, %H:%M:%S')}

🎯 **AÇÃO NECESSÁRIA:**
Claude, por favor responda esta mensagem do WhatsApp usando Write tool para salvar em:
`{CLAUDE_RESPONSE_PATH}`

Formato: `{{"reply": "sua resposta aqui"}}`

Esta é uma mensagem REAL de um usuário do WhatsApp aguardando sua resposta!"""

            # Criar entrada no histórico como se o usuário tivesse digitado
            history_entry = {
                "display": whatsapp_message,
                "pastedContents": {}
            }
            
            # Verificar se o projeto existe no config
            if 'projects' not in claude_config:
                claude_config['projects'] = {}
            
            project_path = '/home/user'
            if project_path not in claude_config['projects']:
                claude_config['projects'][project_path] = {
                    "allowedTools": [],
                    "history": [],
                    "mcpContextUris": [],
                    "mcpServers": {},
                    "enabledMcpjsonServers": [],
                    "disabledMcpjsonServers": [],
                    "hasTrustDialogAccepted": False,
                    "projectOnboardingSeenCount": 1,
                    "hasClaudeMdExternalIncludesApproved": False,
                    "hasClaudeMdExternalIncludesWarningShown": False,
                    "lastTotalWebSearchRequests": 0
                }
            
            # Adicionar no INÍCIO do histórico (mais recente primeiro)
            claude_config['projects'][project_path]['history'].insert(0, history_entry)
            
            # Manter apenas os últimos 100 itens para performance
            if len(claude_config['projects'][project_path]['history']) > 100:
                claude_config['projects'][project_path]['history'] = \
                    claude_config['projects'][project_path]['history'][:100]
            
            # Salvar configuração atualizada
            with open(CLAUDE_CONFIG_PATH, 'w', encoding='utf-8') as f:
                json.dump(claude_config, f, indent=2, ensure_ascii=False)
            
            print(f"✅ Mensagem INJETADA no Claude Config!")
            print(f"🎯 Claude deve ver automaticamente na próxima atualização")
            
            return True
            
        except Exception as error:
            print(f"❌ Erro ao injetar no Claude Config: {error}")
            return False

    def wait_for_claude_response(self, message_id, max_attempts=120):
        """Aguarda resposta do Claude por até 2 minutos"""
        attempts = 0
        
        while attempts < max_attempts and self.running:
            try:
                if os.path.exists(CLAUDE_RESPONSE_PATH):
                    with open(CLAUDE_RESPONSE_PATH, 'r', encoding='utf-8') as f:
                        response_data = json.load(f)
                    
                    if 'reply' in response_data and response_data['reply']:
                        print(f"🎉 CLAUDE RESPONDEU: \"{response_data['reply']}\"")
                        
                        # Limpar arquivo de resposta
                        os.remove(CLAUDE_RESPONSE_PATH)
                        
                        return response_data['reply']
                
                attempts += 1
                time.sleep(1)
                
            except Exception as error:
                print(f"❌ Erro ao verificar resposta: {error}")
                attempts += 1
                time.sleep(1)
        
        print(f"⏰ Timeout - Claude não respondeu em {max_attempts} segundos")
        return None

    def monitor_whatsapp_messages(self):
        """Monitora mensagens do WhatsApp em loop"""
        while self.running:
            try:
                # Verificar se existe arquivo de mensagem
                if not os.path.exists(WHATSAPP_MESSAGES_PATH):
                    time.sleep(2)
                    continue
                
                # Ler mensagem
                with open(WHATSAPP_MESSAGES_PATH, 'r', encoding='utf-8') as f:
                    data = json.load(f)
                
                message = data.get('currentMessage')
                
                # Verificar se é uma nova mensagem
                if not message or message.get('id') == self.last_processed_id:
                    time.sleep(2)
                    continue
                
                print(f"🔔 NOVA MENSAGEM WHATSAPP DETECTADA!")
                print(f"📱 De: {message['senderId']}")
                print(f"💬 Mensagem: \"{message['message']}\"")
                print(f"⏰ Timestamp: {message['timestamp']}")
                
                # Injetar no histórico do Claude
                success = self.inject_message_to_claude_history(
                    message['senderId'], 
                    message['message'], 
                    message['id']
                )
                
                if success:
                    print(f"💉 MENSAGEM INJETADA COM SUCESSO!")
                    print(f"🤖 Claude deve ver automaticamente agora!")
                    print(f"⏳ Aguardando resposta do Claude...")
                    
                    # Aguardar resposta em thread separada para não bloquear
                    response_thread = threading.Thread(
                        target=self.wait_for_claude_response,
                        args=(message['id'],)
                    )
                    response_thread.start()
                
                print(f"========================\n")
                
                # Marcar como processada
                self.last_processed_id = message['id']
                
            except Exception as error:
                print(f"❌ Erro ao processar mensagem WhatsApp: {error}")
                time.sleep(2)

    def start_monitoring(self):
        """Inicia monitoramento em thread separada"""
        monitor_thread = threading.Thread(target=self.monitor_whatsapp_messages)
        monitor_thread.daemon = True
        monitor_thread.start()
        return monitor_thread

# Criar instância global
bridge = ClaudeBridge()

@app.route('/api/whatsapp-chat', methods=['POST'])
def whatsapp_chat():
    """Endpoint para receber mensagens do WhatsApp"""
    try:
        data = request.get_json()
        sender_id = data.get('senderId')
        message = data.get('message')
        
        print(f"\n🔔 NOVA MENSAGEM VIA API!")
        print(f"📱 De: {sender_id}")
        print(f"💬 Mensagem: \"{message}\"")
        print(f"⏰ Horário: {datetime.now().strftime('%d/%m/%Y, %H:%M:%S')}")
        
        # Processar mensagem diretamente
        message_id = str(int(time.time() * 1000))
        
        # Injetar no histórico do Claude
        success = bridge.inject_message_to_claude_history(sender_id, message, message_id)
        
        if success:
            print(f"💉 Mensagem injetada! Aguardando Claude...")
            
            # Aguardar resposta do Claude
            reply = bridge.wait_for_claude_response(message_id)
            
            if reply:
                print(f"📤 Enviando resposta para WhatsApp: \"{reply}\"")
                return jsonify({"reply": reply})
            else:
                fallback_reply = "Desculpe, não consegui processar sua mensagem no momento. Tente novamente."
                return jsonify({"reply": fallback_reply})
        else:
            return jsonify({"reply": "Erro interno do servidor. Tente novamente."})
        
    except Exception as error:
        print(f"❌ Erro na API: {error}")
        return jsonify({"reply": "Erro interno do servidor."}), 500

@app.route('/api/test', methods=['GET'])
def test():
    """Endpoint de teste"""
    return jsonify({
        "status": "Claude Bridge Python ativo!",
        "timestamp": datetime.now().isoformat(),
        "claude_config_exists": os.path.exists(CLAUDE_CONFIG_PATH),
        "message": "Sistema pronto para integração com Claude Code"
    })

@app.route('/api/status', methods=['GET'])
def status():
    """Status do sistema"""
    return jsonify({
        "bridge_running": bridge.running,
        "last_processed_id": bridge.last_processed_id,
        "files_status": {
            "whatsapp_messages": os.path.exists(WHATSAPP_MESSAGES_PATH),
            "claude_config": os.path.exists(CLAUDE_CONFIG_PATH),
            "claude_response": os.path.exists(CLAUDE_RESPONSE_PATH)
        }
    })

if __name__ == '__main__':
    print("🔄 Iniciando monitoramento automático...")
    bridge.start_monitoring()
    
    print("🌐 Iniciando servidor Flask...")
    print("🚀 Claude Bridge Python completo ativo!")
    print("📋 Endpoints disponíveis:")
    print("   • POST /api/whatsapp-chat - Receber mensagens")
    print("   • GET /api/test - Teste do sistema")
    print("   • GET /api/status - Status detalhado")
    print("✍️ Claude deve usar Write tool para responder!\n")
    
    app.run(host='0.0.0.0', port=3001, debug=False)